### DISCUSSION TITLE
Translating Exceptions into Problem Details Responses

### DISCUSSION THREAD

**Discussion by giscus** [timdeschryver/timdeschryver.dev#273] [Announcements]
# Translating Exceptions into Problem Details Responses

In this post, we'll take a look at the newly introduced IExceptionHandler in ASP.NET Core 8. We'll implement an exception handler that translates exceptions into Problem Details using the Problem Details Service. The result is a standardized and better experience for your API consumers.

https://timdeschryver.dev/blog/translating-exceptions-into-problem-details-responses

**Comment by adescalzo** [comment]
Typo? in https://timdeschryver.dev/blog/translating-exceptions-into-problem-details-responses#problem-details 

`IServiceCollection.AddProblemDetails()` instead of `IServiceCollection.UseProblemDetails()`

thanks for the post

**Reply by timdeschryver** [reply]
Good catch, you're right.

**Comment by wbuck** [comment]
I'm curious, above you're adding the `instance` key in the extensions section of the problem details:

```C#
AddProblemDetails(options =>
        options.CustomizeProblemDetails = ctx =>
        {
            ctx.ProblemDetails.Extensions.Add("instance", $"{ctx.HttpContext.Request.Method} {ctx.HttpContext.Request.Path}");
        });
```

Shouldn't it be added to the `Instance` property instead?

```C#
builder.Services.AddProblemDetails(options => options.CustomizeProblemDetails = ctx =>
{
    ctx.ProblemDetails.Instance = $"{ctx.HttpContext.Request.Method} {ctx.HttpContext.Request.Path}";
});
```

Not trying to nit, just curious if there was a reason you added it to the `extensions` instead of the `instance`.

**Comment by zwoolli** [comment]
Thank you for your article! Very helpful.

I have used the `IProblemDetailsWriter` to make a custom writer for unauthorized requests. In my scenario, I needed the `detail` property to contain a message. I do not think creating an `IExceptionHandler` would address this scenario since an unauthorized request does not throw an exception. 

I am curious if I should have injected the `IProblemDetailsService` rather than the `IProblemDetailsFactory`, as you did. 

```
public class UnauthorizedProblemDetailsWriter : IProblemDetailsWriter
{
    private readonly ProblemDetailsFactory _problemDetailsFactory;
    public UnauthorizedProblemDetailsWriter(ProblemDetailsFactory problemDetailsFactory)
    {
        _problemDetailsFactory = problemDetailsFactory;  
    }

    public bool CanWrite(ProblemDetailsContext context)
    {
        return context.HttpContext.Response.StatusCode == StatusCodes.Status401Unauthorized;
    }

    public ValueTask WriteAsync(ProblemDetailsContext context)
    {
        // Create message to user
        string detail = "Please log in to perform this action.";

        // Recreating the problem details to get all customizations from the factory
        ProblemDetails problemDetails = _problemDetailsFactory.CreateProblemDetails
        (
            context.HttpContext,
            context.ProblemDetails.Status ?? context.HttpContext.Response.StatusCode,
            context.ProblemDetails.Title,
            context.ProblemDetails.Type,
            detail,
            context.ProblemDetails.Instance
        );

        if (context.ProblemDetails?.Extensions is not null)
        {
            foreach (var extension in context.ProblemDetails.Extensions)
            {
                problemDetails.Extensions[extension.Key] = extension.Value;
            }
        }
        
        return new ValueTask(context.HttpContext.Response.WriteAsJsonAsync(problemDetails));
    }
}
```

**Comment by neilkonitzermarathon** [comment]
I notice that in TryHandleAsync, the HttpContext.Session is always null. In using this in an MVC application, how would one be able to access any session state parameters to add to logging?

**Comment by valbers** [comment]
`type` should be a URI (see your own description) and not the exception's name.
